enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,330 @@
+
+	// tokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// templates allow you to customize pages like the login page.
+	// +optional
+	Templates OAuthTemplates `json:"templates"`
+	// audit specifies what should be audited in the context of OAuthServer.
+	// +optional
+	// +kubebuilder:default:={"profile":"WriteLoginEvents"}
+	Audit OAuthAudit `json:"audit"`
+}
+
+// OAuthAudit specifies the Audit profile in use.
+type OAuthAudit struct {
+	// profile is a simple drop in profile type that can be turned off by
+	// setting it to "None" or it can be turned on by setting it to
+	// "WriteLoginEvents".
+	// +kubebuilder:default:="WriteLoginEvents"
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
+}
+
+// OAuthAuditProfileType defines a simple audit profile, which can turn OAuth
+// authentication audit logging on or off.
+// +kubebuilder:validation:Enum=None;WriteLoginEvents
+type OAuthAuditProfileType string
+
+const (
+	// "None" disables audit logs.
+	OAuthNoneAuditProfileType AuditProfileType = "None"
+
+	// "WriteLoginEvents" logs login and login failure events.
+	// This is the default.
+	OAuthWriteLoginEventsProfileType AuditProfileType = "WriteLoginEvents"
+)
+```
+
+### User Stories
+
+#### As an administrator I want to inspect successful login and login failure attempts
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect oauth-server audit logs.
+
+#### As an administrator I want to inspect logout events and the deleted user
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect apiserver audit logs for deleted `oauthaccesstoken` resources.
+Audit log entries include the deleted user.
+
+#### As an administrator I want to inspect the source IPs of the client triggering an authentication event
+
+### API Extensions
+
+We introduce a new property called `audit` to the `OAuth` CRD.
+
+This enhancement is not expected to affect any other team or compoment outside of auth.
+
+### Operational Aspects of API Extensions
+
+There is no explicit operational involvement expected. The property is set by default and optional.
+
+It could be turned of by operational interaction with the API extension.
+
+#### Failure Modes
+
+It is expected that all components work properly with the property set, not set or set with bogus data.
+
+#### Support Procedures
+
+If the `audit` can't be set in the CR, support would need to file a bugzilla.
+If there are no audit events for login, login failures or logout events, verify that the `audit` values is set in the `oauth` CR.
+If the `audit` is set to `profile: "WriteLoginEvents"` and there are no audit events for the oauth-server after the login, login failure or logout events, support would need to file a bugzilla.
+
+### Risks and Mitigations
+
+There are no big risks. As with logging, we could always log something that we shouldn't and violate compliance.
```

Comment:
nit: "no known big risks".

Reply:
actually, there are risks:
- DoS like situations where we have a lot of incoming requests issuing many audit logging events, see https://github.com/openshift/enhancements/pull/894#discussion_r707216430 for details.
- Disk capacity protection: audit log rotation, please find the bits in the apiserver audit logging framework.